src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -694,13 +689,44 @@ impl RustcTargetData {
             target_config.insert(ct, host_config.clone());
         }
 
-        Ok(RustcTargetData {
+        let mut res = RustcTargetData {
             rustc,
+            config,
+            requested_kinds: requested_kinds.into(),
             target_config,
             target_info,
             host_config,
             host_info,
-        })
+        };
+
+        // Get all kinds we currently know about. Further kinds will
+        // be added via the `add_kind` function below.
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| {
+                p.manifest()
+                    .default_kind()
+                    .into_iter()
+                    .chain(p.manifest().forced_kind())
+            }));
+        for kind in all_kinds {
+            res.add_kind(&kind)?;
+        }
+
+        Ok(res)
+    }
+
+    pub fn add_kind(&mut self, kind: &CompileKind) -> CargoResult<()> {
```

Comment:
Does this need to be a public method? It seems like it could probably be inlined above perhaps?

Reply:
The reason why it's `pub` is because it will make it easy to add kinds after-the-fact: just call `add_kind` at the places the kinds are added and that should be it.

Inlining in the above would certainly be possible — as far as I remember it's basically the way it was done before, and I've extracted this function in the process of adding kinds everywhere required before noticing that as we don't have artifact dependencies yet and there was thus no place to call this function from yet.

Do you want me to remove this function, or should I let it available so that it's easier to implement artifact dependencies when the time comes?

Reply:
For now let's go ahead and leave it folded into the constructor above, we can always easily add it back later if necessary.